src/nvim/ops.c

```diff
@@ -1681,43 +1681,56 @@ int op_replace(oparg_T *oap, int c)
       /* Compute bytes needed, move character count to num_chars. */
       num_chars = numc;
       numc *= (*mb_char2len)(c);
-      /* oldlen includes textlen, so don't double count */
-      n += numc - bd.textlen;
 
       oldp = get_cursor_line_ptr();
-      oldlen = STRLEN(oldp);
-      assert(n >= 0);
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      oldlen = (int)STRLEN(oldp);
+
+      size_t newp_size = (size_t)(bd.textcol + bd.startspaces);
+      if (had_ctrl_v_cr || (c != '\r' && c != '\n')) {
+        newp_size += (size_t)numc;
+        if (!bd.is_short) {
+          newp_size += (size_t)(bd.endspaces + oldlen
+                                - bd.textcol - bd.textlen + 1);
```

Comment:
The `+1` here is for the terminating `NUL`. We need it always not just in this case. To add it, we could move the `+1` to the declaration of `newp_size` or remove it and use `newp = xmallocz(newp_size);`.

Problem can be seen with `5ia<C-v>u1234<cr><esc>ggl4jr<cr>` and ASAN
```c
READ of size 2 at 0x60200002d531 thread T0
    #0 0x464379 in __interceptor_strlen.part.25 /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:227
    #1 0xe483f4 in ml_flush_line /home/oni-link/git/neovim/src/nvim/memline.c:2647:26
    #2 0xe433e2 in ml_append /home/oni-link/git/neovim/src/nvim/memline.c:1836:5
    #3 0x10a6e65 in op_replace /home/oni-link/git/neovim/src/nvim/ops.c:1733:9
    #4 0xfa9924 in do_pending_operator /home/oni-link/git/neovim/src/nvim/normal.c:1965:9
    #5 0x10694f3 in normal_finish_command /home/oni-link/git/neovim/src/nvim/normal.c:890:3
    #6 0xfdfd6c in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #7 0x166f3d0 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:56:26
    #8 0xf94260 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #9 0xd55611 in main /home/oni-link/git/neovim/src/nvim/main.c:541:3
    #10 0x7fb5483095af in __libc_start_main (/lib64/libc.so.6+0x205af)
    #11 0x444f08 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x444f08)

0x60200002d531 is located 0 bytes to the right of 1-byte region [0x60200002d530,0x60200002d531)
allocated by thread T0 here:
    #0 0x4f66d0 in calloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:72
    #1 0xe786a2 in xcalloc /home/oni-link/git/neovim/src/nvim/memory.c:119:15
    #2 0x10a5759 in op_replace /home/oni-link/git/neovim/src/nvim/ops.c:1696:14
    #3 0xfa9924 in do_pending_operator /home/oni-link/git/neovim/src/nvim/normal.c:1965:9
    #4 0x10694f3 in normal_finish_command /home/oni-link/git/neovim/src/nvim/normal.c:890:3
    #5 0xfdfd6c in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #6 0x166f3d0 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:56:26
    #7 0xf94260 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #8 0xd55611 in main /home/oni-link/git/neovim/src/nvim/main.c:541:3
    #9 0x7fb5483095af in __libc_start_main (/lib64/libc.so.6+0x205af)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:227 in __interceptor_strlen.part.25
```